src/journal/journald-context.c

```diff
@@ -252,9 +252,10 @@ static int client_context_read_cgroup(Server *s, ClientContext *c, const char *u
 
         /* Try to acquire the current cgroup path */
         r = cg_pid_get_path_shifted(c->pid, s->cgroup_root, &t);
-        if (r < 0) {
+        if (r < 0 || streq_ptr(t, "/")) {
```

Comment:
I think this check is insufficient since it doesn't account for `CLONE_NEWCGROUP`. You probably want something like `manager_owns_host_root_cgroup`.

Reply:
We don't support systems really where processes reside in the top-level cgroup (it's not permitted on cgroupsv2 anyway...), hence I think the test @msekletar added is actually OK, though there really should be a comment added here that both explains why we filter this out, and why it's safe to do so.

Reply:
also, please use empty_or_root() for this check

Reply:
> (it's not permitted on cgroupsv2 anyway...)

Hmm, what do you mean by this? Processes at root cgroup are permitted in cgroup v2 (and what that means is controller-specific)

Reply:
OK, true cgroupsv2 has some weird rules about the host root cgroup. But in general in the systemd view of the world userspace processes cannot exist in the root cgroup. That's because it is necessarily an inner node of the cgroup tree (as PID 1 moves itself to /init.scope early on), and inner nodes cannot have processes. And the kernel enforces that at least in a cgroup namespace, and we for all the other cases too.

Reply:
> > 
> 
> But there is a comment already that explains the exception for zombie process under v1.

Hmm, the comment says nothing about the connection between "process is a zombie" and "process is in the root cgroup", that's what I was looking for though, as it's not an obvious rule.

Reply:
/* We use the unit ID passed in as fallback if we have nothing cached yet and cg_pid_get_path_shifted()
  * failed or process is running in a root cgroup. Zombie processes are automatically migrated to root cgroup
  * on cgroupsv1 and we want to be able to map log messages from them too.

Reply:
sounds good to me